In [ ]:
import pandas as pd

In [ ]:
#1. Імпортуй датасет у Pandas
# шлях до датасету на гугл диску
file_id = "12pYkauOit0YEfNocUfc4DgRGcyuiIXH6"
url = f"https://drive.google.com/uc?id={file_id}"

df = pd.read_csv(url)

#https://drive.google.com/file/d/12pYkauOit0YEfNocUfc4DgRGcyuiIXH6/view?usp=sharing

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

In [ ]:
#2. Приведи Date до типу datetime
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')
#перевірка рядків, які не конвертувалися
print("Дати, які не конвертувалися:")
print(df[df['Date'].isna()])

#імпортуємо оригінальний датасет
df_in = pd.read_csv(url)
#окремо обробляємо дату з відмінним форматом
df.loc[26, 'Date'] = pd.to_datetime(df_in.loc[26, 'Date'], format='%Y%m%d', errors='coerce')
print("\nЗначення в рядку 26 після виправлення:")
print(df.loc[26, 'Date'])
print("\nТип даних стовпця Date:")
print(df['Date'].dtype)

Дати, які не конвертувалися:
    Duration Date  Pulse  Maxpulse  Calories
22        45  NaT    100       119     282.0

Значення в рядку 26 після виправлення:
2020-12-26 00:00:00

Тип даних стовпця Date:
datetime64[ns]


Після спроби приведення формату колонки до типу datetime, не конвертувалося два значення (рядки 22 та 26). У рядку 22 - порожнє значення, у рядку 26 - формат дати (YYYYMMDD) не відповідає формату ('YYYY/MM/DD') у інших рядках. Його було оброблено окремим кодом, значення виправлено.

Інший спосіб конвертувати дату:

In [ ]:
df["Date"] = pd.to_datetime(df["Date"], format="mixed", errors="coerce")
print(df)

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

In [ ]:
#3. Перевір структуру таблиці: shape, columns, dtypes, info(), head()
print (f"Розмір таблиці: {df.shape}")
print(f"\nМістить колонки: {df.columns}")
print(f"\nЗ наступними типами:\n{df.dtypes}")
print("\nОгляд структури:")
df.info()
print("\nПерші 5 рядків:")
df.head()

Розмір таблиці: (32, 6)

Містить колонки: Index(['Duration', 'Date', 'Pulse', 'Maxpulse', 'Calories',
       'is_calories_missing'],
      dtype='object')

З наступними типами:
Duration                        int64
Date                   datetime64[ns]
Pulse                           int64
Maxpulse                        int64
Calories                      float64
is_calories_missing              bool
dtype: object

Огляд структури:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Duration             32 non-null     int64         
 1   Date                 31 non-null     datetime64[ns]
 2   Pulse                32 non-null     int64         
 3   Maxpulse             32 non-null     int64         
 4   Calories             32 non-null     float64       
 5   is_calories_missing  32 non-null     bool          
dtype

,Duration,Date,Pulse,Maxpulse,Calories,is_calories_missing
0,60,2020-12-01,110,130,409.1,False
1,60,2020-12-02,117,145,479.0,False
2,60,2020-12-03,103,135,340.0,False
3,45,2020-12-04,109,175,282.4,False
4,45,2020-12-05,117,148,406.0,False


In [ ]:
#4. Перевір наявність пропусків, виведи частку
missing_info = pd.DataFrame({
    'Кількість пропусків': df.isna().sum(),
    'Частка пропусків (%)': df.isna().mean()*100
})

print(missing_info)


          Кількість пропусків  Частка пропусків (%)
Duration                    0                 0.000
Date                        1                 3.125
Pulse                       0                 0.000
Maxpulse                    0                 0.000
Calories                    2                 6.250


In [ ]:
#5. Розрахуй статистики для Calories: mean, median, min, max, std, var
print(f"Середнє значення для Calories: {df['Calories'].mean()}")
print(f"Медіана для Calories: {df['Calories'].median()}")
print(f"Мінімальне значення для Calories: {df['Calories'].min()}")
print(f"Максимальне значення для Calories: {df['Calories'].max()}")
print(f"Стандартне відхилення для Calories: {df['Calories'].std()}")
print(f"Дисперсія для Calories: {df['Calories'].var()}")

Середнє значення для Calories: 304.68
Медіана для Calories: 291.2
Мінімальне значення для Calories: 195.1
Максимальне значення для Calories: 479.0
Стандартне відхилення для Calories: 66.00377941111951
Дисперсія для Calories: 4356.498896551724


Середнє значення 304.68 трохи вище за медіану 291.2
Це вказує на те, що розподіл спалених калорій є злегка скошеним вправо (тобто є кілька тренувань із дуже високим показником калорій)

Діапазон значень досить широкий (від 195.1 до 479.0), а стандартне відхилення (66.00) показує помірну варіативність: більшість тренувань мають показники, що відхиляються від середнього приблизно на +/- 66 калорій.

In [ ]:
#6. Розрахуй Q1, Q3, IQR і знайди викиди
# Розраховуємо квартилі
q1 = df['Calories'].quantile(0.25)
q3 = df['Calories'].quantile(0.75)
print(f"Q1 = {q1}")
print(f"Q3 = {q3}")
# Міжквартильний розмах
iqr = q3 - q1
print(f"IQR = {iqr}")
# Знаходимо межі викидів
low = q1 - 1.5*iqr
high = q3 + 1.5*iqr
print(f"Нижня межа: {low}")
print(f"Верхня межа: {high}")
print(f"Нормальний діапазон: {low:.0f} - {high:.0f} калорій")

# Знаходимо всі викиди
outliers = df[(df["Calories"] < low) | (df["Calories"] > high)]

print(f"Кількість викидів: {outliers.shape[0]}")
print(f"Відсоток викидів: {(outliers.shape[0] / df.shape[0]) * 100:.1f}%")


Q1 = 250.7
Q3 = 343.975
IQR = 93.27500000000003
Нижня межа: 110.78749999999994
Верхня межа: 483.88750000000005
Нормальний діапазон: 111 - 484 калорій
Кількість викидів: 0
Відсоток викидів: 0.0%


Перевірка даних показала, що у стовпці Calories немає жодного викиду, оскільки жодне значення не є меншим за 111 і не є більшим за 484.

In [ ]:
#7. Створи колонку is_calories_missing, заповни пропуски середнім значенням
#створюємо нову колонку із булевим значенням True/False в залежності від наявності пропуску
df['is_calories_missing'] = df['Calories'].isna()
#заповнюємо пропуски у Calories середнім значенням
df['Calories'].fillna(df['Calories'].mean(), inplace = True)
#виводимо результат
print(df[['Calories', 'is_calories_missing']])

    Calories  is_calories_missing
0     409.10                False
1     479.00                False
2     340.00                False
3     282.40                False
4     406.00                False
5     300.00                False
6     374.00                False
7     253.30                False
8     195.10                False
9     269.00                False
10    329.30                False
11    250.70                False
12    250.70                False
13    345.30                False
14    379.30                False
15    275.00                False
16    215.20                False
17    300.00                False
18    304.68                 True
19    323.00                False
20    243.00                False
21    364.20                False
22    282.00                False
23    300.00                False
24    246.00                False
25    334.50                False
26    250.00                False
27    241.00                False
28    304.68  

/tmp/ipython-input-2154471284.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Calories'].fillna(df['Calories'].mean(), inplace = True)


Після заповнення пропусків у Calories середнім значенням, попередній код покаже, що у нас з'явився 1 викид

In [ ]:
#8. Створи колонку pulse_range = Maxpulse - Pulse
df['pulse_range'] = df['Maxpulse'] - df['Pulse']
#виводимо результат
print(df[['Maxpulse', 'Pulse', 'pulse_range']])

    Maxpulse  Pulse  pulse_range
0        130    110           20
1        145    117           28
2        135    103           32
3        175    109           66
4        148    117           31
5        127    102           25
6        136    110           26
7        134    104           30
8        133    109           24
9        124     98           26
10       147    103           44
11       120    100           20
12       120    100           20
13       128    106           22
14       132    104           28
15       123     98           25
16       120     98           22
17       120    100           20
18       112     90           22
19       123    103           20
20       125     97           28
21       131    108           23
22       119    100           19
23       101    130          -29
24       132    105           27
25       126    102           24
26       120    100           20
27       118     92           26
28       132    103           29
29       1

In [ ]:
#9. Створи категорії за Duration: <30 30–60 >60
for i in range(0,31,1):
  if df.loc[i, 'Duration'] < 30:
    df.loc[i, 'Duration_Category'] = '<30'
  elif df.loc[i, 'Duration'] <= 60:
    df.loc[i, 'Duration_Category'] = '30-60'
  else:
    df.loc[i, 'Duration_Category'] = '>60'
#виводимо результат
print(df[['Duration', 'Duration_Category']])


    Duration Duration_Category
0         60             30-60
1         60             30-60
2         60             30-60
3         45             30-60
4         45             30-60
5         60             30-60
6         60             30-60
7        450               >60
8         30             30-60
9         60             30-60
10        60             30-60
11        60             30-60
12        60             30-60
13        60             30-60
14        60             30-60
15        60             30-60
16        60             30-60
17        60             30-60
18        45             30-60
19        60             30-60
20        45             30-60
21        60             30-60
22        45             30-60
23        60             30-60
24        45             30-60
25        60             30-60
26        60             30-60
27        60             30-60
28        60             30-60
29        60             30-60
30        60             30-60
31      

In [ ]:
#10. Побудуй агреговану таблицю з розрахунком середнього Pulse і Calories для кожної групи
duration_info = df.groupby('Duration_Category').agg({
    'Pulse' : 'mean',
    'Calories' : 'mean'
}).reset_index()

duration_info.columns = ['duration category', 'avg_pulse', 'avg_calories']
print(duration_info)

  duration category   avg_pulse  avg_calories
0             30-60  103.866667    308.717857
1               >60  104.000000    253.300000
